src/include/catalog/catalog_cache.h

```diff
@@ -0,0 +1,104 @@
+ * CatalogAccessor will transparently look in a cache first if it exists. If the cache is passed in as nullptr, then the
+ * CatalogAccessor performs its lookup from the DatabaseCatalog as normal. Most operations are expected to only be
+ * performed by CatalogAccessor, which is why most of this class is private and the CatalogAccessor is designated as a
+ * friend class.
+ *
+ * Right now it only caches lookups for table and index pointers, and indexes on a table. In the future if we see other
+ * DatabaseCatalog requests proving expensive, we can add them to the cache.
+ */
+class CatalogCache {
+ public:
+  /**
+   * Clear the cache, and stash a timestamp associated with the start time of the transaction that cleared it. This acts
+   * as a watermark for when this snapshot of the DatabaseCatalog is valid.
+   * @param now start time of the TransactionContext performing the reset
+   */
+  void Reset(const transaction::timestamp_t now) {
+    oldest_entry_ = now;
+    pointers_.clear();
+    indexes_.clear();
+  }
+
+  /**
+   * @return The oldest possible age of an entry in this cache, corresponding to the last cache reset
+   */
+  transaction::timestamp_t OldestEntry() const { return oldest_entry_; }
+
+ private:
+  common::ManagedPointer<storage::SqlTable> GetTable(const table_oid_t table) {
+    const auto key = static_cast<uint32_t>(table);
+    const auto it = pointers_.find(key);
+    if (it != pointers_.end()) {
+      const auto value = it->second;
+      return common::ManagedPointer(reinterpret_cast<storage::SqlTable *>(value));
+    }
+    return nullptr;
+  }
+
+  void PutTable(const table_oid_t table, const common::ManagedPointer<storage::SqlTable> table_ptr) {
+    const auto key = static_cast<uint32_t>(table);
+    const auto value = reinterpret_cast<uintptr_t>(table_ptr.Get());
+    TERRIER_ASSERT(pointers_.count(key) == 0, "Shouldn't be inserting something that already exists.");
+    pointers_[key] = value;
+  }
+
+  common::ManagedPointer<storage::index::Index> GetIndex(const index_oid_t index) {
+    const auto key = static_cast<uint32_t>(index);
+    const auto it = pointers_.find(key);
+    if (it != pointers_.end()) {
+      const auto value = it->second;
+      return common::ManagedPointer(reinterpret_cast<storage::index::Index *>(value));
+    }
+    return nullptr;
+  }
+
+  void PutIndex(const index_oid_t index, const common::ManagedPointer<storage::index::Index> index_ptr) {
+    const auto key = static_cast<uint32_t>(index);
+    const auto value = reinterpret_cast<uintptr_t>(index_ptr.Get());
+    TERRIER_ASSERT(pointers_.count(key) == 0, "Shouldn't be inserting something that already exists.");
+    pointers_[key] = value;
+  }
+
+  std::pair<bool, std::vector<index_oid_t>> GetIndexOids(const table_oid_t table) {
+    const auto it = indexes_.find(table);
+    if (it != indexes_.end()) {
+      return {true, it->second};
+    }
+    return {false, {}};
```

Comment:
Can you add a comment that the bool is added to differentiate from a table with no indexes